fem/gridfunc.cpp

```diff
@@ -2476,35 +2517,76 @@ void GridFunction::ProjectCoefficient(VectorCoefficient &vcoeff)
          }
          SetSubVector(vdofs, vals);
       }
+   }
+   else
+   {
+      constexpr real_t signal = std::numeric_limits<real_t>::min();
+      for (int i = 0; i < fes->GetNE(); i++)
+      {
+         doftrans = fes->GetElementVDofs(i, vdofs);
+         vals.SetSize(vdofs.Size());
+         vals = signal;
+         fes->GetFE(i)->Project(vcoeff, *fes->GetElementTransformation(i), vals);
+         if (doftrans)
+         {
+            doftrans->TransformPrimal(vals);
+         }
+         // Remove undefined dofs
+         // The knot location (either Botella, Demko or Greville point)
+         // where the NURBS dof are evaluated might fall outside of the
+         // domain of the element. In that case the value is not set, and
+         // the value remains the signal value.
+         int s = 0;
+         for (int ii = 0; ii < vals.Size(); ii++)
+         {
+            if (vals[ii] != signal)
+            {
+               vdofs[s] = vdofs[ii];
+               vals(s) = vals(ii);
+               s++;
+            }
+         }
+         vdofs.SetSize(s);
+         vals.SetSize(s);
 
+         // Add reduced dofs to global vector
+         SetSubVector(vdofs, vals);
+      }
    }
+}
 
-   else
+void GridFunction::ProjectCoefficientGlobalL2(VectorCoefficient &vcoeff,
+                                              real_t rtol, int iter)
+{
+   // Define and assemble linear form
+   LinearForm b(fes);
+   BilinearForm a(fes);
+   if (fes->FEColl()->GetRangeType(3)  == mfem::FiniteElement::VECTOR)
```

Comment:
As the `GetRange` is asked to the `FiniteElementCollection` as a whole and not a particular `Element` a dimension needs to be provided. I chose 3, but this could also be 2, that is immaterial from this perspective.
Could this create issues on the other side? I think we should cross that bridge when we are there...
For now I will add a comment that this dimension is chosen arbitrarily.

Reply:
Thanks for documenting. I thought this dimension should come from the `Mesh::Dimension()`, but I think you are right that 2 or 3 should work.